Invoices.elm

```diff
@@ -24,45 +25,50 @@ type alias Product =
     , unit : Int
     , tax : Float
     , price : Float
-    , total : Float
     }
 
 
 type alias Model =
-    { total : Float
-    , products : List Product
+    { products : List Product
     }
 
 
 initModel : Model
 initModel =
-    Model 0 []
+    Model []
 
 
-createProduct : Int -> Product
-createProduct id =
-    Product id "" 1 21 0.0 0.0
+addNew : Model -> Model
+addNew { products } =
+    products
+        ++ [ { id = List.length products, item = "", unit = 1, tax = 21, price = 0 } ]
+        |> Model
 
 
 calculateProductTotal : Product -> Float
 calculateProductTotal product =
-    (calculateProductPriceTotal product) + (calculateProductTax product)
+    calculateProductPriceTotal product + calculateProductTax product
 
 
 calculateProductPriceTotal : Product -> Float
-calculateProductPriceTotal product =
-    let
-        unit =
-            toFloat product.unit
-    in
-        unit * product.price
+calculateProductPriceTotal { unit, price } =
+    toFloat unit * price
 
 
 calculateProductTax : Product -> Float
-calculateProductTax product =
-    (calculateProductPriceTotal product) * (product.tax / 100)
+calculateProductTax ({ tax } as product) =
+    calculateProductPriceTotal product * (tax / 100)
+
+
+updateWithId : id -> ({ a | id : id } -> { a | id : id }) -> List { a | id : id } -> List { a | id : id }
+updateWithId id predicate =
+    List.map (\p -> if id == p.id then predicate p else p)
 
 
+updateProducts : (List Product -> List Product) -> Model -> Model
+updateProducts updateWith ({ products } as model) =
+    { model | products = updateWith products}
```

Comment:
Frankly, I don't know if I would invent this in some bigger app but there was not much space to show how more things can be build up by stacking small functions together.